### PR TITLE
Polynomial symbolic cost

### PR DESCRIPTION
**Pull request by hongkai-dai**
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/5661)

### CODE REVIEW

**Comment by hongkai-dai** [comment]
+@russtedrake for feature review please. Interestingly github also recommends you to review.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-KgRQoq5ARheoYHuxnhU:b-ogvu9q)*

**Comment by soonho-tri** [comment]
+@soonho-tri for platform review.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-KgRQwtqwXZXKuqdD4Xk:bxw8cpy)*

**Comment by hongkai-dai** [comment]
+(status: do not review) I have some test failures, that are solver related. Some solver (IPOPT) performs very differently from another solver (SNOPT)

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-KgRbPHxZluy1OImcsPx:b-g6ci0v)*

**Comment by hongkai-dai** [comment]
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgSClMpoLsIQ-K69yYU:-KgSClMqt6ZllfFtPU16:b-fconek) ([raw file](https://github.com/robotlocomotion/drake/blob/92570c2c8e19f76d09087a8fdae641e28cbd5664/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L62)):*
> ```cpp
>   }
> 
>   // Test interpolation.
> ```

There is something I do not understand, on why we test interpolation in this way 
1. Why the linear interpolation HAS TO have a different mid point, compared to the spline interpolation?
2. If they are different, the difference depend on the length of the time interval `t2 - t1`. By adding the running cost that depends on the time interval, the optimal solution has a shorter time interval than before, and thus the difference gets smaller.
3. Why we choose `times_out[2]` and `times_out[3]`, instead of other knot point?
@lucy-tri could you shed some light on this test? Thanks a lot!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by hongkai-dai** [comment]
-(status: do not review) 

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-KgSKeqkGxvVst0acZDu:ba0k9zt)*

**Comment by hongkai-dai** [comment]
@russtedrake and @soonho-tri . The PR is ready to review.

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-KgSTc566eJbe7WPArAU:b6ayycy)*

**Comment by Lucy-tri** [comment]
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgSClMpoLsIQ-K69yYU:-KgUKaw5Bhhq4o_P1vRo:b-clfkjn) ([raw file](https://github.com/robotlocomotion/drake/blob/92570c2c8e19f76d09087a8fdae641e28cbd5664/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L62)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

There is something I do not understand, on why we test interpolation in this way 
1. Why the linear interpolation HAS TO have a different mid point, compared to the spline interpolation?
2. If they are different, the difference depend on the length of the time interval `t2 - t1`. By adding the running cost that depends on the time interval, the optimal solution has a shorter time interval than before, and thus the difference gets smaller.
3. Why we choose `times_out[2]` and `times_out[3]`, instead of other knot point?
@lucy-tri could you shed some light on this test? Thanks a lot!
</blockquote></details>

I picked those points 2 and 3 because I found experimentally that the cubic  interpolation between those points was different from the linear interpolation, for that data set. I was trying to prove that the cubic was actually different than linear (making sure that the correct interpolation had been chosen). I added this test after we corrected the dircol traj to actually do spline interpolation, I think (one of the last features we added to traj opt). I didn't see anywhere else to test this new code, at the top level.  It's possible there is a better way to do this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by Lucy-tri** [comment]
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgSClMpoLsIQ-K69yYU:-KgULWNfY-7_9VhYRJ-B:bkc4vm4) ([raw file](https://github.com/robotlocomotion/drake/blob/92570c2c8e19f76d09087a8fdae641e28cbd5664/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L62)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

I picked those points 2 and 3 because I found experimentally that the cubic  interpolation between those points was different from the linear interpolation, for that data set. I was trying to prove that the cubic was actually different than linear (making sure that the correct interpolation had been chosen). I added this test after we corrected the dircol traj to actually do spline interpolation, I think (one of the last features we added to traj opt). I didn't see anywhere else to test this new code, at the top level.  It's possible there is a better way to do this. 
</blockquote></details>

In other words, we couldn't see any other way to actually test that cubic was being used in pendulum, at the time. We wanted just the most basic kind of go/no-go test possible.  I see now you need to delete this test. Ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by hongkai-dai** [comment]
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgSClMpoLsIQ-K69yYU:-KgUgwwXqvmuoSpKoZ2e:b-umc1rv) ([raw file](https://github.com/robotlocomotion/drake/blob/92570c2c8e19f76d09087a8fdae641e28cbd5664/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L62)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

In other words, we couldn't see any other way to actually test that cubic was being used in pendulum, at the time. We wanted just the most basic kind of go/no-go test possible.  I see now you need to delete this test. Ok.
</blockquote></details>

I see, thanks for the explanation. If the goal is to test the cubic interpolation, would it be better to compute the midpoint from the cubic interpolation formula by hand (with x[n], x[n+1], xdot[n], xdot[n+1], we can solve the 4 parameters of the cubic function, and then compute the midpoint), then compare this midpoint value against the `spline_interpolation` you computed? Doing it in this way, we are guaranteed that the code uses mid point interpolation, and the behavior does not depend on the exact vale of `state_traj`. I think the previous test depends too much on the exact value of `state_traj`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by Lucy-tri** [comment]
Reviewed 1 of 10 files at r1.
Review status: 1 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgSClMpoLsIQ-K69yYU:-KgUmYRe56d6fU7K7b9q:b-qy3f03) ([raw file](https://github.com/robotlocomotion/drake/blob/92570c2c8e19f76d09087a8fdae641e28cbd5664/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L62)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I see, thanks for the explanation. If the goal is to test the cubic interpolation, would it be better to compute the midpoint from the cubic interpolation formula by hand (with x[n], x[n+1], xdot[n], xdot[n+1], we can solve the 4 parameters of the cubic function, and then compute the midpoint), then compare this midpoint value against the `spline_interpolation` you computed? Doing it in this way, we are guaranteed that the code uses mid point interpolation, and the behavior does not depend on the exact vale of `state_traj`. I think the previous test depends too much on the exact value of `state_traj`.
</blockquote></details>

Your suggestion sounds great, and it would prove that cubic interpolation is being used (my test merely proved that it was not linear interpolation).
BTW,  if the interpolation is really working well with a nice spline, the number of time samples should be able to be reduced (from 21), and still be able to swing up to the goal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by hongkai-dai** [comment]
Review status: 1 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgSClMpoLsIQ-K69yYU:-KgV0t7P5p1h5VsOrz0P:b5j53um) ([raw file](https://github.com/robotlocomotion/drake/blob/92570c2c8e19f76d09087a8fdae641e28cbd5664/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L62)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

Your suggestion sounds great, and it would prove that cubic interpolation is being used (my test merely proved that it was not linear interpolation).
BTW,  if the interpolation is really working well with a nice spline, the number of time samples should be able to be reduced (from 21), and still be able to swing up to the goal.
</blockquote></details>

I just added the test we discussed.

We can check if 21 is the best number in a separate PR (It is out of the scope of this PR). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by Lucy-tri** [comment]
Reviewed 3 of 10 files at r1, 1 of 1 files at r2.
Review status: 5 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgSClMpoLsIQ-K69yYU:-KgVSw0HlohkW13CyXr1:bb7k0qh) ([raw file](https://github.com/robotlocomotion/drake/blob/92570c2c8e19f76d09087a8fdae641e28cbd5664/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L62)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I just added the test we discussed.

We can check if 21 is the best number in a separate PR (It is out of the scope of this PR). 
</blockquote></details>

lgtm

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 75 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgVSWStV-ogT7Uht6jp:-KgVSWStV-ogT7Uht6jq:b-k9byy4) ([raw file](https://github.com/robotlocomotion/drake/blob/f6e824656c5edc1bfc0726100a17fb9d7dd0e59f/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L75)):*
> ```cpp
>     // xdot_l, xdot_r, and thus the midpoint of the polynomial can be computed
>     // from x_l, x_r, xdot_l, xdot_r. Check out equation 9 of
>     // Direct Trajectory Optimization Using NOnlinear Programming and
> ```

NIT: NOnlinear should be Nonlinear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by soonho-tri** [comment]
Reviewed 7 of 10 files at r1.
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by soonho-tri** [comment]
@hongkai-dai , is it closed by accident?

**Comment by hongkai-dai** [comment]
I was stupid enough to accidentally delete this branch on the remote. Just restored the branch. Thanks!

**Comment by hongkai-dai** [comment]
@drake-jenkins-bot linux-xenial-gcc-ninja-experimental please

---

Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-Kg_yBiNL4Qg5SxxZMAb:bko8goz)*

**Comment by hongkai-dai** [comment]
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 75 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgVSWStV-ogT7Uht6jp:-Kg_yOoKc9vthk2fEOok:byoeemj) ([raw file](https://github.com/robotlocomotion/drake/blob/f6e824656c5edc1bfc0726100a17fb9d7dd0e59f/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L75)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

NIT: NOnlinear should be Nonlinear.
</blockquote></details>

Thanks! I will change this when the other reviewers make their comments, and then send the change altogether.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by RussTedrake** [comment]
one minor request, otherwise :lgtm:

---

Reviewed 8 of 10 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgiNX7QTcDhrLPn2yWe:-KgiNX7R5cXw5vG-fxZ2:b-yybe1s) ([raw file](https://github.com/robotlocomotion/drake/blob/f6e824656c5edc1bfc0726100a17fb9d7dd0e59f/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L80)):*
> ```cpp
>     const Eigen::Vector2d spline_midpoint(state_traj.value((t_l + t_r) / 2));
>     EXPECT_TRUE(CompareMatrices(xm, spline_midpoint, 1E-10,
>                                 MatrixCompareType::absolute));
> ```

BTW, Great to see this test.  thanks.

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 117 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgiPHr30LCSqSbaXdln:-KgiPHr30LCSqSbaXdlo:bv71hvl) ([raw file](https://github.com/robotlocomotion/drake/blob/f6e824656c5edc1bfc0726100a17fb9d7dd0e59f/drake/systems/trajectory_optimization/direct_collocation.cc#L117)):*
> ```cpp
>   //    sum_{i=0...N-2} h_i/2.0 * (g_i + g_{i+1}), or
>   // g_0*h_0/2.0 + [sum_{i=1...N-2} g_i*(h_{i-1} + h_i)/2.0] +
>   // g_{N-1}*h_{N-2}/2.0.
> ```

BTW, oops.  thanks!

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 125 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgiPeVUAjUHKHStKjUW:-KgiPeVUAjUHKHStKjUX:buj4pfu) ([raw file](https://github.com/robotlocomotion/drake/blob/f6e824656c5edc1bfc0726100a17fb9d7dd0e59f/drake/systems/trajectory_optimization/direct_collocation.cc#L125)):*
> ```cpp
>   }
>   AddCost(0.5 *
>           SubstitutePlaceholderVariables(g * h_vars()(N() - 2) / 2, N() - 1));
> ```

could you add a trivial traj opt test that checks that the integrated cost is what we expect?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-KgiPpxXe0TZC8Y8RXkK:b-y4zdn3)*

**Comment by hongkai-dai** [comment]
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_collocation.cc, line 125 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgiPeVUAjUHKHStKjUW:-Kgjfb8xfCxKudM2Bwjt:b-q7lijj) ([raw file](https://github.com/robotlocomotion/drake/blob/f6e824656c5edc1bfc0726100a17fb9d7dd0e59f/drake/systems/trajectory_optimization/direct_collocation.cc#L125)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

could you add a trivial traj opt test that checks that the integrated cost is what we expect?
</blockquote></details>

Good call! I think when #5685 is merged, it will be easy to add that test? I added a TODO in `pendulum_trajectory_optimization_test.cc`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by hongkai-dai** [comment]
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc, line 75 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-KgVSWStV-ogT7Uht6jp:-Kgjh8IkymFdJ9ipnWYk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f6e824656c5edc1bfc0726100a17fb9d7dd0e59f/drake/examples/Pendulum/test/pendulum_trajectory_optimization_test.cc#L75)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Thanks! I will change this when the other reviewers make their comments, and then send the change altogether.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by soonho-tri** [comment]
Reviewed 1 of 1 files at r2.
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661)*

**Comment by RussTedrake** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-KgjiLUs_dh_ANIKCZTo:bnfp4nl)*

**Comment by soonho-tri** [comment]
:lgtm:

---

Reviewed 2 of 10 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5661#-:-Kgjk0DeET_jm2v5uI12:bnfp4nl)*
